### DISCUSSION TITLE
Any suggestions for more sophisticated EOT handling than simply firing a python assert?

### DISCUSSION THREAD

**Discussion by apaiva** [pyuvm/pyuvm#224] [Q&A]
I'm looking to enable some capabilities to EOT handling beyond a simple immediate fail on a python assert.  Some features of interest:
- Allow N errors before actually terminating the test
- Call back into a component phase (final?) before terminating so they can report state.  (Particularly interesting on timeouts when you want to know what's outstanding.)

Has there been any conversation on home grown solutions or building this into pyuvm?

**Comment by raysalemi** [comment]
Thanks for the question!

It's easy to program the first behavior by catching the assertion.  In fact, it's good practice to catch the assertions and then drop the objection in the run phase when you reach your error limit.

Dropping the objection in the run phase allows pyuvm to move on to the `extract_phase()` and ` report_phase()`, which implement your second request.  You just override these phases in your base test to get the behavior you want.
